src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepExecution.java

```diff
@@ -230,29 +239,41 @@ public String getShortDescription() {
         }
     }
 
-    public interface ResetCallback extends Serializable {
-        void logWritten();
-    }
+    private static final class ResetTimer extends SlaveToMasterCallable<Void, RuntimeException> {
 
-    private class ResetCallbackImpl implements ResetCallback {
         private static final long serialVersionUID = 1L;
-        @Override public void logWritten() {
-            resetTimer();
+
+        private final String id;
+
+        ResetTimer(String id) {
+            this.id = id;
         }
+
+        @Override public Void call() throws RuntimeException {
+            StepExecution.applyAll(TimeoutStepExecution.class, e -> {
+                if (id.equals(e.id)) {
+                    e.resetTimer();
+                }
+                return null;
+            });
+            return null;
+        }
+
     }
 
     private static class ConsoleLogFilterImpl extends ConsoleLogFilter implements /* TODO Remotable */ Serializable {
         private static final long serialVersionUID = 1L;
 
-        private final ResetCallback callback;
+        private final @Nonnull String id;
+        private transient @CheckForNull Channel channel;
 
-        ConsoleLogFilterImpl(ResetCallback callback) {
-            this.callback = callback;
+        ConsoleLogFilterImpl(String id) {
+            this.id = id;
         }
 
-        private Object writeReplace() {
-            Channel ch = Channel.current();
-            return ch == null ? this : new ConsoleLogFilterImpl(ch.export(ResetCallback.class, callback));
```

Comment:
Maybe the issue is that only the marshaller holds the reference to the exported object, so after the object is completely serialized it is free to be garbage collected?

Reply:
AFAICT the issue was that [this code](https://github.com/jenkinsci/remoting/blob/81d1beda0dd2cbc6ca845d1bc812e45c553ae980/src/main/java/hudson/remoting/UserRequest.java#L299) released all exported objects when a `Callable` finished, insofar as those objects had originally been exported within the dynamic scope of that `Callable`’s serialization. Or something like this. I could not follow exactly what the logic was, and the documentation was scant. At any rate, the revised strategy with the `SlaveToMasterCallable` is simpler to reason about explicitly.